apps/emqx/src/bpapi/emqx_bpapi_trans.erl

```diff
@@ -0,0 +1,195 @@
+                                      , casts => Casts
+                                      }),
+    {function, Line, ?META_FUN, _Arity = 0,
+     [{clause, Line, _Args = [], _Guards = [],
+       [Ret]}]}.
+
+mk_export() ->
+    {attribute, 0, export, [{?META_FUN, 0}]}.
+
+is_attribute({attribute, _Line, _Attr, _Val}) -> true;
+is_attribute(_)                               -> false.
+
+%% Extract the target function of the RPC call
+analyze_fun(Line, Name, Arity, [{clause, Line, Head, _Guards, Exprs}], S) ->
+    analyze_exprs(Line, Name, Arity, Head, Exprs, S);
+analyze_fun(Line, Name, Arity, _Clauses, S) ->
+    invalid_fun(Line, Name, Arity, S).
+
+analyze_exprs(Line, Name, Arity, Head, Exprs, S) ->
+    log("~p/~p (~p):~n~p", [Name, Arity, Head, Exprs]),
+    try
+        [{call, _, CallToBackend, CallArgs}] = Exprs,
+        OuterArgs = extract_outer_args(Head),
+        Key = {S#s.module, Name, OuterArgs},
+        {Semantics, Target} = extract_target_call(CallToBackend, CallArgs),
+        push_target({Semantics, Key, Target}, S)
+    catch
+        _:Err:Stack ->
+            log("Failed to process function call:~n~s~nStack: ~p", [Err, Stack]),
+            invalid_fun(Line, Name, Arity, S)
+    end.
+
+-spec extract_outer_args([erl_parse:abstract_form()]) -> [atom()].
+extract_outer_args(Abs) ->
+    lists:map(fun({var, _, Var}) ->
+                      Var;
+                 ({match, _, {var, _, Var}, _}) ->
+                      Var;
+                 ({match, _, _, {var, _, Var}}) ->
+                      Var
+              end,
+              Abs).
+
+-spec extract_target_call(_AST, [_AST]) -> {semantics(), emqx_bpapi:call()}.
+extract_target_call(RPCBackend, OuterArgs) ->
+    {Semantics, {atom, _, M}, {atom, _, F}, A} = extract_mfa(RPCBackend, OuterArgs),
+    {Semantics, {M, F, list_to_args(A)}}.
+
+-define(BACKEND(MOD, FUN), {remote, _, {atom, _, MOD}, {atom, _, FUN}}).
+-define(IS_RPC(MOD), (MOD =:= erpc orelse MOD =:= rpc)).
+
+%% gen_rpc:
+extract_mfa(?BACKEND(gen_rpc, _), _) ->
+    %% gen_rpc has an extremely messy API, thankfully it's fully wrapped
+    %% by emqx_rpc, so we simply forbid direct calls to it:
+    error("direct call to gen_rpc");
+%% emqx_rpc:
+extract_mfa(?BACKEND(emqx_rpc, CallOrCast), [_Node, M, F, A]) ->
+    {call_or_cast(CallOrCast), M, F, A};
+extract_mfa(?BACKEND(emqx_rpc, CallOrCast), [_Tag, _Node, M, F, A]) ->
+    {call_or_cast(CallOrCast), M, F, A};
+%% (e)rpc:
+extract_mfa(?BACKEND(RPC, CallOrCast), [_Node, M, F, A]) when ?IS_RPC(RPC) ->
+    {call_or_cast(CallOrCast), M, F, A};
+extract_mfa(?BACKEND(RPC, CallOrCast), [_Node, M, F, A, _Timeout]) when ?IS_RPC(RPC) ->
+    {call_or_cast(CallOrCast), M, F, A};
+extract_mfa(_, _) ->
+    error("unrecognized RPC call").
+
+call_or_cast(cast)      -> cast;
+call_or_cast(multicast) -> cast;
+call_or_cast(multicall) -> call;
+call_or_cast(call)      -> call.
+
+list_to_args({cons, _, {var, _, A}, T}) ->
```

Comment:
Small suggestion: since this only matches pure variables (like `call(Topic)` but not `call({key, Topic})`), would it make sense to add a fallback clause that adds something like `Arg3` if the argument is not a simple variable?

Reply:
>  if the argument is not a simple variable
This case is not supported. It should just crash and result in this https://github.com/emqx/emqx/pull/6629/files#diff-048bea61123ebd1af9e3cbfee089790dc190d5d701c2effc86995bcf86f3a675R40 errro

Reply:
I thought that would be a legitimate use case (to call with a tuple, for example).  Is it related to the type checking?

Reply:
The idea is that decorator functions are trivial and transparent. If some preprocessing is needed, it can be done outside, in the module invoking the bpapi function..